weld/optimizer/transforms/cse.rs

```diff
@@ -0,0 +1,573 @@
+                    let replacement = Expr {
+                        ty: ty,
+                        kind: Ident(name),
+                        annotations: Annotations::new(),
+                    };
+                    Some(replacement)
+                }
+            }
+        });
+    }
+
+    /// Builds a scope map for each identifier.
+    ///
+    /// The scope map maps a symbol to the *least deep* (i.e., highest) scope in an expression tree.
+    fn build_scope_map(&mut self,
+                       expr: &Expr,
+                       bindings: &HashMap<Symbol, Expr>) -> HashMap<Symbol, ScopeDepth> {
+        let mut scopes = HashMap::new();
+        // Start at -1 so the top-level lambda gets a depth of 0.
+        self.build_scope_map_helper(expr, bindings, &mut scopes, -1);
+        scopes
+    }
+
+    fn build_scope_map_helper(&mut self,
+                           expr: &Expr,
+                           bindings: &HashMap<Symbol, Expr>,
+                           scope_map: &mut HashMap<Symbol, ScopeDepth>,
+                           current_scope: ScopeDepth) {
+
+        let mut handled = true;
+        match expr.kind {
+            If { ref cond, ref on_true, ref on_false } => {
+                self.build_scope_map_helper(cond, bindings, scope_map, current_scope);
+                // Update current scope for branch targets.
+                self.build_scope_map_helper(on_true, bindings, scope_map, current_scope + 1);
+                self.build_scope_map_helper(on_false, bindings, scope_map, current_scope + 1);
+            }
+            Lambda { ref body, .. } => {
+                self.build_scope_map_helper(body, bindings, scope_map, current_scope + 1);
+            }
+            Ident(ref name) if bindings.contains_key(name) => {
+                {
+                    let ent = scope_map.entry(name.clone()).or_insert(current_scope);
+                    if current_scope < *ent {
+                        *ent = current_scope;
+                    }
+                }
+                let expr = bindings.get(name).unwrap();
+                self.build_scope_map_helper(expr, bindings, scope_map, current_scope);
+            }
+            _ => {
+                handled = false;
+            }
+        }
+
+        if !handled {
+            for child in expr.children() {
+                self.build_scope_map_helper(child, bindings, scope_map, current_scope);
+            }
+        }
+    }
+
+    /// Generates bindings in the expression.
+    ///
+    /// This injects the definition of each symbol into the correct position in the final
+    /// expression. Specifically, a symbol is defined in the highest scope that symbol is used in.
+    ///
+    /// The function updates `expr` with the symbol to value mappings in `bindings`. Bindings that
+    /// are already inserted are tracked in `generated`. The `stack` tracks scopes to generate
+    /// the bindings in, and `scopes` maps each symbol to its highest scope.
```

Comment:
I would add an explicit statement about what the scopes in the stack mean (list of (symbol, expression) mappings) as opposed to what's in the `scopes` map (which is a symbol to `scope_id` mapping)